go-controller/pkg/ovn/controller/egress_services/egress_services_controller.go

```diff
@@ -417,16 +490,338 @@ func (c *Controller) repair() error {
+
+	state := c.services[key]
+	if es == nil {
+		if state == nil {
+			// The EgressService was deleted and wasn't configured before.
+			// We delete it from the unallocated service cache just in case.
+			delete(c.unallocatedServices, key)
+			return nil
+		}
+		// The service is configured but does no longer have the egress service resource,
+		// meaning we should clear all of its resources.
+		return c.clearServiceResources(key, state)
+	}
+
+	if svc == nil {
+		if state == nil {
+			// The service object was deleted and was not an allocated egress service.
+			// We delete it from the unallocated service cache just in case.
+			delete(c.unallocatedServices, key)
+			return c.setEgressServiceHost(namespace, name, "")
+		}
+		// The service was deleted and was an egress service.
+		// We delete all of its relevant resources to avoid leaving stale configuration.
+		return c.clearServiceResources(key, state)
+	}
+
+	// At this point both the EgressService resource and the Service != nil
+
+	if state != nil && state.stale {
+		// The service is marked stale because something failed when trying to delete it.
+		// We try to delete it again before doing anything else.
+		return c.clearServiceResources(key, state)
+	}
+
+	if state == nil && len(svc.Status.LoadBalancer.Ingress) == 0 {
+		// The service wasn't configured before and does not have an ingress ip.
+		// we don't need to configure it and make sure it does not have a stale host value or unallocated entry.
+		delete(c.unallocatedServices, key)
+		return c.setEgressServiceHost(namespace, name, "")
+	}
+
+	if state != nil && len(svc.Status.LoadBalancer.Ingress) == 0 {
+		// The service has no ingress ips so it is not considered valid anymore.
+		return c.clearServiceResources(key, state)
+	}
+
+	nodeSelector := es.Spec.NodeSelector
+	v4Endpoints, v6Endpoints, epsNodes, err := c.allEndpointsFor(svc)
+	if err != nil {
+		return err
+	}
+
+	// If the service is ETP=Local we'd like to add an additional constraint to the selector
+	// that only a node with local eps can be selected. Otherwise new ingress traffic will break.
+	if len(epsNodes) != 0 && svc.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyTypeLocal {
+		matchEpsNodes := metav1.LabelSelectorRequirement{
+			Key:      "kubernetes.io/hostname",
+			Operator: metav1.LabelSelectorOpIn,
+			Values:   epsNodes,
+		}
+		nodeSelector.MatchExpressions = append(nodeSelector.MatchExpressions, matchEpsNodes)
+	}
+	selector, err := metav1.LabelSelectorAsSelector(&nodeSelector)
+	if err != nil {
+		return err
+	}
+	totalEps := len(v4Endpoints) + len(v6Endpoints)
+
+	// We don't want to select a node for a service without endpoints to not "waste" an
+	// allocation on a node.
+	if totalEps == 0 && state == nil {
+		c.unallocatedServices[key] = selector
+		return c.setEgressServiceHost(namespace, name, "")
+	}
+
+	if totalEps == 0 && state != nil {
+		c.unallocatedServices[key] = selector
+		return c.clearServiceResources(key, state)
```

Comment:
lets assume clearServiceResources fails due to some txn error. Now services map still has the key for this service in it, but also unallocatedServices has the key. Is that OK or should we not set unallocated until its actually been deallocated.

Reply:
it is okay, since txn error = state stale, and once it is cleared it will either get allocated, or stay in the unallocated cache.